utils/rowgroup/rowstorage.cpp

```diff
@@ -666,6 +666,156 @@ class RowGroupStorage
+        }
+        return true;
+      }
+
+      if (tgid >= fFinalizedRows.size())
+        fFinalizedRows.resize(tgid + 1, 0ULL);
+
+      // Check if there are rows to process
+      bool hasReturnRows = false;
+      for (auto i = fgid; i < tgid; ++i)
+      {
+        if (fFinalizedRows[i] != ~0ULL)
+        {
+          // Not all rows are finalized, we have to return at least parts of this RGData
+          hasReturnRows = true;
+          break;
+        }
+      }
+
+      if (rgdata)
+      {
+        // RGData is currently in memory
+        if (!hasReturnRows)
+        {
+          // All rows are finalized, don't return this RGData
+          continue;
+        }
+      }
+      else
+      {
+        if (hasReturnRows)
+        {
+          // Load RGData from disk, unlink dump file and continue processing
+          loadRG(rgid, rgdata, true);
+        }
+        else
+        {
+          // All rows are finalized. Unlink dump file and continue search for return RGData
+          unlink(makeRGFilename(rgid).c_str());
+          continue;
+        }
+      }
+
+      // TODO: Finish comments (make statements out of assumptions)
+      // assumption: this shifts data within RGData such that it compacts the non finalized rows
+      // move next non-finalized rows on finalized row positions
+      //
+      uint64_t pos = 0;
+      uint64_t opos = 0;
+      fRowGroupOut->setData(rgdata.get());
+      for (auto i = fgid; i < tgid; ++i)
+      {
+        if ((i - fgid) * 64 >= fRowGroupOut->getRowCount())
+          break;
+        uint64_t mask = ~fFinalizedRows[i];
+        if ((i - fgid + 1) * 64 > fRowGroupOut->getRowCount())
+        {
+          mask &= (~0ULL) >> ((i - fgid + 1) * 64 - fRowGroupOut->getRowCount());
+        }
+        opos = (i - fgid) * 64;
+
+        if (mask == ~0ULL)
+        {
+          if (LIKELY(pos != opos))
+            moveRows(rgdata.get(), pos, opos, 64);
+          pos += 64;
+          continue;
+        }
+
+        if (mask == 0)
+          continue;
+
+        while (mask != 0)
+        {
+          // assmp.: find position until block full of not finalized rows?
+          size_t b = __builtin_ffsll(mask);
+          size_t e = __builtin_ffsll(~(mask >> b)) + b;
+          if (UNLIKELY(e >= 64))
+            mask = 0;
+          else
+            mask >>= e;
+          if (LIKELY(pos != opos + b - 1))
+            moveRows(rgdata.get(), pos, opos + b - 1, e - b);
+          pos += e - b;
+          opos += e;
+        }
+        --opos;
+      }
+
+      // assmp.:: nothing got shifted at all -> all rows must be finalized? (but shouldn't it be the
+      // other way around?) if all rows finalized remove RGData and file and don't give it out why would
+      // we check this here again? see: fRowGroupOut->setRowCount(pos); -> pos is the number of valid
+      // rows, interesting
+      if (pos == 0)
+      {
+        fLRU->remove(rgid);
+        unlink(makeRGFilename(rgid).c_str());
```

Comment:
probably unlinks should be done in RAII manner

Reply:
Thank you for your comment! What do you mean by RAll manner? (Sorry, I am probably missing vocabulary here.)

Reply:
RAII in this case means we call a hypotetical `AggTempFile` struct ctor that going out of scope call `unlink` in its dtor.